### DISCUSSION TITLE
When including font file in build output, add `<link rel="preload">` for those files

### DISCUSSION THREAD

**Discussion by jkjustjoshing** [vercel/next.js#36699] [Ideas]
### Describe the feature you'd like to request

When Next.js finds a `@font-face` statement in the CSS that points to a local font file, Next.js puts that file into the `.next/static/media` folder with a content hash. However, the font files don't have a `<link rel="preload">` element to start loading them before the CSS is downloaded and parsed. This creates a request waterfall that prevents the page's text from displaying as quickly as it could.

Next.js should provide some way to preload necessary font files.

### Describe the solution you'd like

Naive solution: any font file that is encountered and moved to the `.next/static/media` folder should have a corresponding `<link rel="preload>` element.
Issue with naive solution: just because a font is referenced in a `@font-face` rule, doesn't mean that that font is referenced by a CSS statement, and it doesn't mean that CSS statement applies to any text on the page. Preloading every font file could have a negative performance impact due to loading more fonts than the page actually requires.

Better solution: expose the list of fonts found to the page somehow, and allow the developer to determine what to preload, either in `_document.js` or `_app.js`.

### Describe alternatives you've considered

If the font exists in the `/public` folder, then we know the path to it at dev time (because there's no content hash on the outputted file). However, as @alexasselin008 describes in [his discussion post here](https://github.com/vercel/next.js/discussions/35554), this isn't always possible if the `@font-face` rules aren't under our control.

I tried playing around with the `file-loader` Webpack plugin, but I was unable to get the content hash to match the content hash that Next.js creates and injects into the CSS.

**Comment by 7iomka** [comment]
any news on this?
